### PR TITLE
Fix default arguments bug and improve `MethodRedefinitionWarning`

### PR DESCRIPTION
**Pull request by wesselb**
What the title says. WIP.

### CODE REVIEW

**Comment by PhilipVinc** [comment]
@wesselb The warning is very noisy and several users are complaining. Can we revert the previous PR until we have something that works fine?

**Comment by wesselb** [comment]
@PhilipVinc I agree that it's noisy, but it's also already uncovered some issues, so it does have significant utility.

How about an opt-out mechanism, where the warning is supressed if the module of the implementation that overwrites the current method is part of a list?

You'd then do something like

```python
from plum import opt_out_method_redefinition_warning_for_package

opt_out_method_redefinition_warning_for_package("netket")
```

**Comment by PhilipVinc** [comment]
What issues?

File: plum/function.py [hunk 1]
```diff
@@ -417,6 +417,8 @@ def invoke(self, *types: TypeHint) -> Callable:
         def wrapped_method(*args, **kw_args):
             return _convert(method(*args, **kw_args), return_type)
 
+        wrapped_method.__wrapped_by_plum__ = method
```
Comment (PhilipVinc) [inline plum/function.py:420]: Can you add a docstring to explain what this does and why?

Comment (wesselb) [reply plum/function.py:420]: Definitely! Will also add a test which illustrates the edge case. Sorry, haven’t had a moment yet to do this.

Comment (wesselb) [reply plum/function.py:435]: Added in the tests!
